graph2tac/tfgnn/tasks.py

```diff
@@ -1206,6 +1231,37 @@ def _log_softmax_logits(
 
         return local_logits, global_logits
 
+    @staticmethod
+    def transpose_ragged_first_dim(
+        x: tf.RaggedTensor,  # [outer*inner, None(ragged), ...]
+        outer_dim: int,
+        inner_dim: int,
+    ) -> tf.RaggedTensor:  # [inner*outer, None(ragged), ...]
+        """
+        Convert a ragged tensor of shape `[outer*inner, None(ragged), ...]`
+        to one of shape `[inner*outer, None(ragged), ...]`
+        by permuting the inner values.
+        """
+        inner_dim = tf.cast(inner_dim, dtype=tf.int64)
```

Comment:
Tensorflow doesn't support ragged tensors of shape `[a, b, None(c)]`, i.e. where the first two or more dimensions are not ragged.  So I uses few tricks where I combine `a` and `b` into one dimension.  This is a transpose operator for this case.